keps/sig-apps/2255-pod-cost/README.md

```diff
@@ -55,102 +48,129 @@ Items marked with (R) are required *prior to targeting to a milestone / release*
 
-This feature allows making a suggestion to the ReplicaSet controller, which pod of a Deployment should be deleted first when a scale-down event happens. This can prevent session disruption in stateful applications in a trivial manner.
+This feature allows applications to give a hint to the ReplicaSet controller 
+as to which pods should be deleted first on scale down.
 
 ## Motivation
 
-For some applications, it is necessary that the application can tell Kubernetes which pod can be deleted and which replica has to be protected. The reason for this is that some applications do have stateful sessions and it is not possible to put such an application into Kubernetes because of session termination resulting from "random" down-scale. If the application is able to tell Kubernetes which of the replicas contains no/few/less important active sessions, this would solve many problems. This feature is non-disruptive to the default behaviour. Only if the annotation is existing, it will make a difference in deletion order.
+Currently ReplicaSets are scaled down based on a criteria that on the
+limit prioritizes deleting pods with a more recent creation/readiness 
+timestamp. This is not ideal for some applications where the cost of 
+deleting pods is not related to how recent they were created. 
 
 ### Goals
 
-To recommend which pod gets deleted next of a ReplicaSet. This should help to avoid major reworks in existing applications architecture:
-* [45509](https://github.com/kubernetes/kubernetes/issues/45509) - Scale down a deployment by removing specific pods
+- Allow applications to influence the order of deleting pods when scaling down a ReplicaSet
 
 
 ### Non-Goals
 
-Guaranteed (in contrast to the recommendation stated in Goals) deletion of a selected replica.
+- Guarantees on pod deletion order
 
 ## Proposal
 
-The application can set the `controller.kubernetes.io/pod-cost` annotation to a pod through the Kubernetes API. When a downscale event happens, the pod with the lower priority value of the previously set annotation will be deleted first. If one pod of the Deployment has no priority annotation set, it will be treated as the lowest priority.
+Define a known annotation, namely `controller.kubernetes.io/pod-cost` that 
+applications can set to offer a hint on the cost of deleting a pod compared
+to other pods belonging to the same ReplicaSet. 
 
-If all pods have the same priority, there is no difference in the normal pod delete decision behaviour. The same applies if the pod-cost annotation is not used at all.
+The pod-cost range will be from [-MaxInt, MaxInt]. The default value is 0.
+Invalid values (like setting the annotation to string) will be converted to
+the default value.
 
-The pod-cost annotation can be changed during operation, for example, if workload changes or a new master gets elected.
+When scaling down a ReplicaSet, controller-manager will prioritize deleting
+pods with higher pod-cost. Specifically, the rank passed to 
```

Comment:
Test plan mentions:
```
Integration tests to validate that:
 * Replicas with lower pod-cost are deleted before replicas with higher pod-cost
```
whereas my understanding of `controller-manager will prioritize deleting pods with higher pod-cost` is higher pod-cost replicas are deleted before lower pod-cost replicas.

Reply:
thanks for the catch, I updated the text in the proposal. higher-cost = lower chance of being deleted.

Reply:
To clarify this again: 
higher-cost (higher relative value) means lower priority and higher chance to delete
lower-cost (lower relative value) means higher priority, lower chance to delete

Reply:
hmm, this is confusing, the way I was thinking about it is that pod cost is how costly it is to remove the pod. But you are suggesting to define it as the cost of the pod on the cluster.

Reply:
the former definition is easier to implement by controllers: for example, cost could be the number of open stateful sessions.

Reply:
It is ambigous at the moment, so we have to define it.

In my PR for example I've implemented it like I described above: https://github.com/kubernetes/kubernetes/pull/87849
Of course we can also change this.

Reply:
I think we need to choose the semantics that are easier for users to use.

Reply:
lets continue this discussion here: https://github.com/kubernetes/enhancements/pull/2272#discussion_r558481486